File: packages/api-core/src/defaults/errorMapper.ts
import { ErrorMapper, TIMEOUT_BAG_KEY, TransportResult } from "../types";
import { Request } from "../request";
import { Context } from "../context";
import { CanceledError, NetworkError, TimeoutError, UniReqError } from "../errors";

function isAbortError(err: any): boolean {
  return err?.name === "AbortError";
}

export class DefaultErrorMapper implements ErrorMapper {
  map(err: unknown, _req: Request, ctx: Context, _transport?: TransportResult): Error {
    if (err instanceof UniReqError) return err;

    if (ctx.bag.get(TIMEOUT_BAG_KEY) === true) {
      return new TimeoutError("Request timed out", err);
    }

    if (isAbortError(err)) {
      return new CanceledError("Request canceled", err);
    }

    return new NetworkError("Network error", err);
  }
}

File: packages/api-core-python/src/aptx_api_core/_error_mapper.py
from __future__ import annotations

import httpx

from ._context import Context
from ._request import Request
from .errors import HttpError, NetworkError, TimeoutError, UniReqError


class DefaultErrorMapper:
    def map(self, err: Exception, req: Request, ctx: Context) -> UniReqError:
        if isinstance(err, UniReqError):
            return err
        if isinstance(err, httpx.TimeoutException):
            return TimeoutError(str(err) or "Request timed out")
        if isinstance(err, httpx.ConnectError):
            return NetworkError(str(err) or "Connection failed")
        if isinstance(err, httpx.HTTPStatusError):
            return HttpError(
                f"HTTP {err.response.status_code}",
                status=err.response.status_code,
                url=str(err.request.url),
            )
        return UniReqError(str(err))
